src/Microsoft.VisualStudio.Threading.Analyzers.CSharp/VSTHRD110ObserveResultOfAsyncCallsAnalyzer.cs

```diff
@@ -42,6 +42,87 @@ public override void Initialize(AnalysisContext context)
+
+        private static bool IsPublicProperty(ISymbol symbol) =>
+            symbol.DeclaredAccessibility == Accessibility.Public && symbol.Kind == SymbolKind.Property;
+
+        private static bool IsParameterlessNonGenericMethod(IMethodSymbol methodSymbol) =>
+            methodSymbol.Parameters.Length == 0 && !methodSymbol.IsGenericMethod;
+
+        private static bool IsAwaitable(ITypeSymbol returnedSymbol) =>
+            returnedSymbol.Name switch
+            {
+                Types.Task.TypeName
+                    when returnedSymbol.BelongsToNamespace(Types.Task.Namespace) => true,
+
+                Types.ConfiguredTaskAwaitable.TypeName
+                    when returnedSymbol.BelongsToNamespace(Types.ConfiguredTaskAwaitable.Namespace) => true,
+
+                Types.ValueTask.TypeName
+                    when returnedSymbol.BelongsToNamespace(Types.ValueTask.Namespace) => true,
+
+                Types.ConfiguredValueTaskAwaitable.TypeName
+                    when returnedSymbol.BelongsToNamespace(Types.ConfiguredValueTaskAwaitable.Namespace) => true,
+
+                _ => false,
+            };
+
+        private static bool IsCustomAwaitable(ITypeSymbol returnedSymbol)
+        {
+            // has method: public T GetAwaiter()
+            IMethodSymbol? getAwaiterMethod = returnedSymbol.GetMembers("GetAwaiter")
+                .Where(IsPublicMethod)
+                .Cast<IMethodSymbol>()
+                .Where(m => !m.ReturnsVoid && IsParameterlessNonGenericMethod(m))
+                .SingleOrDefault();
+
+            if (getAwaiterMethod is null)
+            {
+                return false;
+            }
+
+            // examine custom awaiter type
+            ITypeSymbol returnType = getAwaiterMethod.ReturnType;
+
+            // implementing: System.Runtime.CompilerServices.INotifyCompletion
+            if (!returnType.AllInterfaces.Any(i =>
+                IsOfType(i, nameof(System.Runtime.CompilerServices.INotifyCompletion), Namespaces.SystemRuntimeCompilerServices)))
+            {
+                return false;
+            }
+
+            // has property: public bool IsCompleted { get; }
+            IPropertySymbol? isCompletedProperty = returnType.GetMembers("IsCompleted")
+                .Where(IsPublicProperty)
+                .Cast<IPropertySymbol>()
+                .Where(p => p.GetMethod != null && IsOfType(p.Type, nameof(Boolean), Namespaces.System))
+                .SingleOrDefault();
+
+            if (isCompletedProperty is null)
+            {
+                return false;
+            }
+
+            // has method: public void GetResult()
+            IMethodSymbol? getResultMethod = returnType.GetMembers("GetResult")
+                .Where(IsPublicMethod)
+                .Cast<IMethodSymbol>()
+                .Where(m => m.ReturnsVoid && IsParameterlessNonGenericMethod(m))
+                .SingleOrDefault();
```

Comment:
All these use of Linq extension methods incur many allocations for every invocation in a compilation. We strive to have minimal impact to the perf of the compilation, so can you please rewrite this such that we avoid the closures, delegates and `IEnumerable`/`IEnumerator` allocations?

Reply:
Sure, I totally understand!